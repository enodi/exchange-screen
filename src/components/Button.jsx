import React from "react";

const Button = props => (
  <button className={`button ${props.color} ${props.size}`}>
    {props.text}
  </button>
);

export default Button;
