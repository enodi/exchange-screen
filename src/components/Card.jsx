import React from "react";

const Card = props => <div className="card-container">{props.children}</div>;

export default Card;
