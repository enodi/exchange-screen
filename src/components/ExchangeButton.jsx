import React from "react";

const ExchangeButton = () => (
  <button id="exchange-button">
    <svg x="0px" y="0px" viewBox="0 0 470.72 470.72" width="35px" height="35px">
      <g>
        <g>
          <g>
            <path
              d="M193.76,337.6l-48,48c-9.937,9.903-17.583,21.864-22.4,35.04V43.36h-32v376.96c-4.79-13.105-12.38-25.008-22.24-34.88    l-48-48L0,358.72l107.36,107.2l107.36-107.36L193.76,337.6z"
              data-original="#000000"
              className="active-path"
              data-old_color="#000000"
              fill="#FFFFFF"
            />
          </g>
        </g>
        <g>
          <g>
            <path
              d="M363.36,4.8L256,112l20.96,20.96l49.92-49.92c8.951-9.056,15.782-19.985,20-32v376.32h32V50.72    c4.311,12.044,11.253,22.976,20.32,32l49.92,49.92l21.6-20.64L363.36,4.8z"
              data-original="#000000"
              className="active-path"
              data-old_color="#000000"
              fill="#FFFFFF"
            />
          </g>
        </g>
      </g>{" "}
    </svg>
    <span className="sr-only">Click the Swap Currencies</span>
  </button>
);

export default ExchangeButton;
