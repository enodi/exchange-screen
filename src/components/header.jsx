import React from "react";
import { NavLink } from "react-router-dom";
import logo from "images/bar.svg";

const Header = () => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <NavLink className="header__title" to="/">
          <h1>Revolut</h1>
          <img src={logo} className="logo" alt="logo" />
        </NavLink>
      </div>
    </div>
  </header>
);

export default Header;
