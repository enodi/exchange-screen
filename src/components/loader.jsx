import React from "react";
import { Row, Icon } from "antd";

const Loader = () => (
  <Row type="flex" justify="center" align="middle" className="full-height">
    <Icon type="loading" className="sec-color-text" style={{ fontSize: 48 }} />
  </Row>
);

export default Loader;
