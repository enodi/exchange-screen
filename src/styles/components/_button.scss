.button {
  border-radius: 30px;
  color: #fff;
  font-size: $font-size-large;
  cursor: pointer;
}

.button.pink {
  background-color: rgb(235, 0, 141);
  border: 1px solid rgb(235, 0, 141);
}

.button.blue {
  background-color: $dark-blue;
  border: 1px solid $dark-blue;
}

.button.large {
  width: calc(100% - 48px);
  height: 60px;
  margin: 24px;
}

.button.medium {
  width: 150px;
  height: 50px;
}
