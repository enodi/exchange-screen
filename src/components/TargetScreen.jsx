import React from "react";

import Card from "./Card";
import Button from "./Button";

const TargetScreen = () => (
  <>
    <Card>
      <h1 className="text">Set Target</h1>
      <form>
        <div className="top-section">
          <div>
            <select className="pocket">
              <option value="gbp">GBP</option>
              <option value="eur">EUR</option>
              <option value="usd">USD</option>
            </select>
            <span>Balance</span>
          </div>
          <input type="number" placeholder="0" required />
        </div>
        <Button text="Set Target" color="pink" size="large" />
      </form>
    </Card>
  </>
);

export default TargetScreen;
