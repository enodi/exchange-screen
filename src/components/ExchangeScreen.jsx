import React from "react";
import { NavLink } from "react-router-dom";
import { connect } from "react-redux";

import Card from "./Card";
import Button from "./Button";
import ExchangeButton from "./ExchangeButton";
import { fetchExchangeRate } from "actions/exchange-rate";
import rateConverter from "helpers/converter";

export class ExchangeScreen extends React.Component {
  state = {
    disabled: true,
    exchange: false,
    source: {
      value: "GBP",
      amount: "",
      error: false,
      pocket: [
        {
          currency: "GBP",
          symbol: "£",
          balance: 1000
        },
        {
          currency: "EUR",
          symbol: "€",
          balance: 700
        },
        {
          currency: "USD",
          symbol: "$",
          balance: 500
        },
        {
          currency: "JPY",
          symbol: "¥",
          balance: 1200
        }
      ]
    },
    target: {
      value: "EUR",
      amount: "",
      error: false,
      pocket: [
        {
          currency: "GBP",
          symbol: "£",
          balance: 1000
        },
        {
          currency: "EUR",
          symbol: "€",
          balance: 700
        },
        {
          currency: "USD",
          symbol: "$",
          balance: 500
        },
        {
          currency: "JPY",
          symbol: "¥",
          balance: 1200
        }
      ]
    }
  };

  getExchangeRates = () => {
    const {
      source: { value: initialSourceCurrency },
      target: { value: initialTargetCurrency }
    } = this.state;
    this.props.getExchangeRate({
      source: initialSourceCurrency,
      target: initialTargetCurrency
    });
  };

  componentDidMount() {
    this.getExchangeRates();
  }

  rateCheckInterval = setInterval(() => {
    this.getExchangeRates();
    clearInterval(this.rateCheckInterval);
  }, 10 * 1000);

  componentDidUpdate(prevProps, prevState) {
    if (prevState !== this.state) {
      this.getExchangeRates();
    }
  }

  sourcePocketChange = event => {
    const source = { ...this.state.source };
    source.value = event.target.value;
    this.setState({ source });
  };

  targetPocketChange = event => {
    const target = { ...this.state.target };
    target.value = event.target.value;
    this.setState({ target });
  };

  sourceInputChange = event => {
    const { target, source } = this.props.exchangeRate;
    const { value } = event.target;
    const sourcePocket = { ...this.state.source };
    const targetPocket = { ...this.state.target };
    const sourcePocketBalance = this.retrievePocketBalance(
      sourcePocket.pocket,
      sourcePocket.value
    );

    sourcePocket.amount = value;
    if (value.length > 0) {
      this.setState(() => ({ disabled: false }));
      this.setState(() => ({ exchange: true }));

      targetPocket.amount =
        sourcePocket.value === "EUR"
          ? rateConverter(value / source, 100)
          : rateConverter(value * target, 100);

      if (targetPocket.amount > sourcePocketBalance.slice(1)) {
        this.setState(() => ({ disabled: true }));
        sourcePocket.error = true;
        this.setState({ source: sourcePocket });
      } else {
        sourcePocket.error = false;
        this.setState({ source: sourcePocket });
      }
    } else {
      this.setState(() => ({ disabled: true }));
      targetPocket.amount = "";
    }

    this.setState({ source: sourcePocket });
    this.setState({ target: targetPocket });
  };

  targetInputChange = event => {
    const { target, source } = this.props.exchangeRate;
    const { value } = event.target;
    const sourcePocket = { ...this.state.source };
    const targetPocket = { ...this.state.target };
    const targetPocketBalance = this.retrievePocketBalance(
      targetPocket.pocket,
      targetPocket.value
    );

    targetPocket.amount = value;
    if (value.length > 0) {
      this.setState(() => ({ disabled: false }));
      this.setState(() => ({ exchange: false }));
      sourcePocket.amount =
        sourcePocket.value === "EUR"
          ? rateConverter(value * source, 100)
          : rateConverter(value / target, 100);

      if (sourcePocket.amount > targetPocketBalance.slice(1)) {
        this.setState(() => ({ disabled: true }));
        targetPocket.error = true;
        this.setState({ target: targetPocket });
      } else {
        targetPocket.error = false;
        this.setState({ target: targetPocket });
      }
    } else {
      this.setState(() => ({ disabled: true }));
      sourcePocket.amount = "";
    }
    this.setState({ target: targetPocket });
    this.setState({ source: sourcePocket });
  };

  handleExchange = event => {
    event.preventDefault();
    const { exchange } = this.state;
    const sourcePocket = { ...this.state.source };
    const targetPocket = { ...this.state.target };
    let sourceArr = [];
    let targetArr = [];

    if (exchange) {
      sourcePocket.pocket.map(item => {
        if (item.currency === sourcePocket.value) {
          item.balance = item.balance - sourcePocket.amount;
        }
        sourceArr.push(item);
        sourcePocket.pocket = sourceArr;
      });
      targetPocket.pocket.map(item => {
        if (item.currency === targetPocket.value) {
          item.balance = item.balance + targetPocket.amount;
        }
        targetArr.push(item);
        targetPocket.pocket = targetArr;
      });

      sourcePocket.amount = "";
      targetPocket.amount = "";
      this.setState({ source: sourcePocket });
      this.setState({ target: targetPocket });
      this.setState(() => ({ disabled: true }));
    } else {
      sourcePocket.pocket.map(item => {
        if (item.currency === sourcePocket.value) {
          item.balance = item.balance + sourcePocket.amount;
        }
        sourceArr.push(item);
        sourcePocket.pocket = sourceArr;
      });
      targetPocket.pocket.map(item => {
        if (item.currency === targetPocket.value) {
          item.balance = item.balance - targetPocket.amount;
        }
        targetArr.push(item);
        targetPocket.pocket = targetArr;
      });

      sourcePocket.amount = "";
      targetPocket.amount = "";
      this.setState({ source: sourcePocket });
      this.setState({ target: targetPocket });
      this.setState(() => ({ disabled: true }));
    }
  };

  retrievePocketSymbol = (pocket, value) => {
    const selectedPocket = pocket.filter(item => item.currency === value);
    return selectedPocket[0].symbol;
  };

  retrievePocketCurrency = pocket => {
    return pocket.map(({ currency }) => currency);
  };

  retrievePocketBalance = (pocket, value) => {
    const itemsInPocket = pocket.filter(item => item.currency === value);
    return `${itemsInPocket[0].symbol}${rateConverter(
      itemsInPocket[0].balance,
      100
    )}`;
  };

  render() {
    const { source, target } = this.props.exchangeRate;
    const {
      source: {
        value: sourceValue,
        pocket: sourcePocket,
        amount: sourceAmount,
        error: sourceError
      },
      target: {
        value: targetValue,
        pocket: targetPocket,
        amount: targetAmount,
        error: targetError
      },
      disabled
    } = this.state;

    return (
      <>
        <Card>
          <h1 className="text">Exchange Screen</h1>
          <form onSubmit={this.handleExchange}>
            <div className="top-section">
              <div>
                <select
                  className="pocket"
                  value={sourceValue}
                  onChange={this.sourcePocketChange}
                >
                  {sourcePocket &&
                    this.retrievePocketCurrency(sourcePocket)
                      .filter(currency => currency !== targetValue)
                      .map((currency, index) => {
                        return (
                          <option
                            value={currency}
                            key={`pocket-currency-${index}`}
                          >
                            {currency}
                          </option>
                        );
                      })}
                </select>
                <span
                  className={`balance ${sourceError ? "error" : ""}`}
                >{`Balance: ${this.retrievePocketBalance(
                  sourcePocket,
                  sourceValue
                )}`}</span>
              </div>
              <input
                type="number"
                placeholder="0.00"
                required
                onChange={this.sourceInputChange}
                value={sourceAmount}
              />
            </div>
            <div className="middle-section">
              <div className="inner-section">
                <ExchangeButton />
                <div className="exchange-rate">
                  <span>
                    {source
                      ? `${this.retrievePocketSymbol(
                          sourcePocket,
                          sourceValue
                        )}${rateConverter(source, 10000)}`
                      : "0"}{" "}
                    ={" "}
                    {target
                      ? `${this.retrievePocketSymbol(
                          targetPocket,
                          targetValue
                        )}${rateConverter(target, 10000)}`
                      : "0"}
                  </span>
                </div>
                <NavLink to="/target">
                  <Button text="Set target" color="blue" size="medium" />
                </NavLink>
              </div>
            </div>
            <div className="bottom-section">
              <div>
                <select
                  className="pocket"
                  value={targetValue}
                  onChange={this.targetPocketChange}
                >
                  {targetPocket &&
                    this.retrievePocketCurrency(targetPocket)
                      .filter(currency => currency !== sourceValue)
                      .map((currency, index) => {
                        return (
                          <option
                            value={currency}
                            key={`pocket-currency-${index}`}
                          >
                            {currency}
                          </option>
                        );
                      })}
                </select>
                <span
                  className={`balance ${targetError ? "error" : ""}`}
                >{`Balance: ${this.retrievePocketBalance(
                  targetPocket,
                  targetValue
                )}`}</span>
              </div>
              <input
                type="number"
                placeholder="0.00"
                required
                onChange={this.targetInputChange}
                value={targetAmount}
              />
            </div>
            <Button
              text="Exchange"
              color="pink"
              size="large"
              disabled={disabled}
            />
          </form>
        </Card>
      </>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  getExchangeRate: data => dispatch(fetchExchangeRate(data))
});

const mapStateToProps = state => ({
  exchangeRate: state.rates
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExchangeScreen);
