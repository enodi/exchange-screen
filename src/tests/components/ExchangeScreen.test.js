import React from "react";
import { shallow } from "enzyme";
import { ExchangeScreen } from "components/ExchangeScreen";
import rateConverter from "helpers/converter";

describe("ExchangeScreen Component", () => {
  let wrapper, getExchangeRate, exchangeRate;
  beforeEach(() => {
    getExchangeRate = jest.fn();
    exchangeRate = { source: 1, target: 1.1492931847 };
    wrapper = shallow(
      <ExchangeScreen
        exchangeRate={exchangeRate}
        getExchangeRate={getExchangeRate}
      />
    );
  });

  it("should render ExchangeScreen component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should contain h1 element with text", () => {
    expect(wrapper.find("h1").text()).toEqual("Exchange Screen");
  });

  it("should contain currencies in each pocket", () => {
    expect(
      wrapper
        .find("select")
        .at(0)
        .prop("value")
    ).toEqual("GBP");
    expect(
      wrapper
        .find("select")
        .at(1)
        .prop("value")
    ).toEqual("EUR");
  });

  it("should contain pocket balance", () => {
    expect(
      wrapper
        .find("span")
        .at(0)
        .text()
    ).toEqual("Balance: £1000");
    expect(
      wrapper
        .find("span")
        .at(2)
        .text()
    ).toEqual("Balance: €700");
  });

  it("should set amount if valid input", () => {
    const value = "23.5";
    const target = rateConverter(value * exchangeRate.target, 100);
    wrapper
      .find("input")
      .at(0)
      .simulate("change", {
        target: { value }
      });
    expect(wrapper.state().source.amount).toEqual(value);
    expect(wrapper.state().target.amount).toEqual(target);
  });

  it("should disable exchange button if amount specified is more than pocket balance", () => {
    const value = "1200";
    wrapper
      .find("input")
      .at(0)
      .simulate("change", {
        target: { value }
      });
    expect(
      wrapper
        .find("span")
        .at(0)
        .hasClass("error")
    ).toEqual(true);
  });

  it("should handle onSubmit", () => {
    wrapper.find("form").simulate("submit", {
      preventDefault: () => {}
    });
    expect(getExchangeRate).toHaveBeenCalled();
  });
});
