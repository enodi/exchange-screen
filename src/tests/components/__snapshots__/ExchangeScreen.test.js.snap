// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeScreen Component should render ExchangeScreen component correctly 1`] = `
<Fragment>
  <Card>
    <h1
      className="text"
    >
      Exchange Screen
    </h1>
    <form
      onSubmit={[Function]}
    >
      <div
        className="top-section"
      >
        <div>
          <select
            className="pocket"
            onChange={[Function]}
            value="GBP"
          >
            <option
              key="pocket-currency-0"
              value="GBP"
            >
              GBP
            </option>
            <option
              key="pocket-currency-1"
              value="USD"
            >
              USD
            </option>
            <option
              key="pocket-currency-2"
              value="JPY"
            >
              JPY
            </option>
          </select>
          <span
            className="balance "
          >
            Balance: £1000
          </span>
        </div>
        <input
          onChange={[Function]}
          placeholder="0.00"
          required={true}
          type="number"
          value=""
        />
      </div>
      <div
        className="middle-section"
      >
        <div
          className="inner-section"
        >
          <ExchangeButton />
          <div
            className="exchange-rate"
          >
            <span>
              £1
               
              =
               
              €1.1493
            </span>
          </div>
          <NavLink
            to="/target"
          >
            <Button
              color="blue"
              size="medium"
              text="Set target"
            />
          </NavLink>
        </div>
      </div>
      <div
        className="bottom-section"
      >
        <div>
          <select
            className="pocket"
            onChange={[Function]}
            value="EUR"
          >
            <option
              key="pocket-currency-0"
              value="EUR"
            >
              EUR
            </option>
            <option
              key="pocket-currency-1"
              value="USD"
            >
              USD
            </option>
            <option
              key="pocket-currency-2"
              value="JPY"
            >
              JPY
            </option>
          </select>
          <span
            className="balance "
          >
            Balance: €700
          </span>
        </div>
        <input
          onChange={[Function]}
          placeholder="0.00"
          required={true}
          type="number"
          value=""
        />
      </div>
      <Button
        color="pink"
        disabled={true}
        size="large"
        text="Exchange"
      />
    </form>
  </Card>
</Fragment>
`;
